x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -0,0 +1,211 @@
+import org.elasticsearch.action.get.GetAction;
+import org.elasticsearch.action.get.GetRequest;
+import org.elasticsearch.action.search.SearchAction;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.common.xcontent.XContentHelper;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.index.query.QueryBuilders;
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.SearchHits;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+import org.elasticsearch.xcontent.XContentType;
+import org.elasticsearch.xpack.core.security.action.profile.Profile;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.security.support.SecurityIndexManager;
+
+import java.io.IOException;
+import java.time.Clock;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.xpack.core.ClientHelper.SECURITY_ORIGIN;
+import static org.elasticsearch.xpack.core.ClientHelper.executeAsyncWithOrigin;
+import static org.elasticsearch.xpack.security.support.SecuritySystemIndices.SECURITY_PROFILE_ALIAS;
+
+public class ProfileService {
+    private static final Logger logger = LogManager.getLogger(ProfileService.class);
+    private static final String DOC_ID_PREFIX = "profile_";
+
+    private final Settings settings;
+    private final Clock clock;
+    private final Client client;
+    private final SecurityIndexManager profileIndex;
+    private final ThreadPool threadPool;
+
+    public ProfileService(Settings settings, Clock clock, Client client, SecurityIndexManager profileIndex, ThreadPool threadPool) {
+        this.settings = settings;
+        this.clock = clock;
+        this.client = client;
+        this.profileIndex = profileIndex;
+        this.threadPool = threadPool;
+    }
+
+    public void getProfile(String uid, @Nullable Set<String> dataKeys, ActionListener<Profile> listener) {
+        getVersionedDocument(uid, ActionListener.wrap(versionedDocument -> {
+            // TODO: replace null with actual domain lookup
+            listener.onResponse(versionedDocument.toProfile(null, dataKeys));
+        }, listener::onFailure));
+    }
+
+    private void getVersionedDocument(String uid, ActionListener<VersionedDocument> listener) {
+        if (maybeHandleIndexStatusIssue(listener)) {
+            return;
+        }
+
+        final GetRequest getRequest = new GetRequest(SECURITY_PROFILE_ALIAS, uidToDocId(uid));
+        try (ThreadContext.StoredContext ignore = client.threadPool().getThreadContext().stashWithOrigin(SECURITY_ORIGIN)) {
```

Comment:
Is this stash required? I don't think so.

Reply:
No it's not because `executeAsyncWithOrigin` also stashes. Thanks